pkg/scheduler/queue/tree_queue.go

```diff
@@ -0,0 +1,282 @@
+// the receiver/parent node name will create a child node of the same name as the parent.
+func (q *TreeQueue) EnqueueBackByPath(childPath QueuePath, v any) error {
+	childQueue, err := q.getOrAddNode(childPath)
+	if err != nil {
+		return err
+	}
+	if childQueue.localQueue.Len()+1 > childQueue.maxQueueLen {
+		return ErrTooManyRequests
+	}
+
+	childQueue.localQueue.PushBack(v)
+	return nil
+}
+
+// EnqueueFrontByPath enqueues an item in the front of the local queue of the node
+// located at a given path through the tree; nodes for the path are created as needed.
+//
+// Max queue length check is skipped; enqueueing to the front is intended only for items
+// which were first enqueued to the back and then dequeued after reaching the front.
+//
+// Re-enqueueing to the front is intended for cases where a queue consumer fails to
+// complete operations on the dequeued item, but failure is not yet final, and the
+// operations should be retried by a subsequent queue consumer.
+//
+// childPath must be relative to the receiver node; providing a QueuePath beginning with
+// the receiver/parent node name will create a child node of the same name as the parent.
+func (q *TreeQueue) EnqueueFrontByPath(childPath QueuePath, v any) error {
+	childQueue, err := q.getOrAddNode(childPath)
+	if err != nil {
+		return err
+	}
+	childQueue.localQueue.PushFront(v)
+	return nil
+}
+
+// getOrAddNode recursively adds tree queue nodes based on given relative child path.
+//
+// childPath must be relative to the receiver node; providing a QueuePath beginning with
+// the receiver/parent node name will create a child node of the same name as the parent.
+func (q *TreeQueue) getOrAddNode(childPath QueuePath) (*TreeQueue, error) {
+	if len(childPath) == 0 {
+		return q, nil
+	}
+
+	var childQueue *TreeQueue
+	var ok bool
+
+	if childQueue, ok = q.childQueueMap[childPath[0]]; !ok {
+		// no child node matches next path segment
+		// create next child before recurring
+		childQueue = NewTreeQueue(childPath[0], q.maxQueueLen)
+
+		// add new child queue to ordered list for round-robining;
+		// in order to maintain round-robin order as nodes are created and deleted,
+		// the new child queue should be inserted directly before the current child
+		// queue index, essentially placing the new node at the end of the line
+		if q.currentChildQueueIndex == localQueueIndex {
+			// special case; cannot slice into childQueueOrder with index -1
+			// place at end of slice, which is the last slot before the local queue slot
+			q.childQueueOrder = append(q.childQueueOrder, childQueue.name)
+		} else {
+			// insert into order behind current child queue index
+			q.childQueueOrder = append(
+				q.childQueueOrder[:q.currentChildQueueIndex],
+				append(
+					[]string{childQueue.name},
+					q.childQueueOrder[q.currentChildQueueIndex:]...,
+				)...,
+			)
+			// update current child queue index to its new place in the expanded slice
+			q.currentChildQueueIndex++
+		}
```

Comment:
[nit] I wonder if this should be its own "insert item at index in slice" method (or if one already exists that we can use here)